### PR TITLE
**New Resource:** `alicloud_ess_eci_scaling_configuration`

### PR DESCRIPTION
**Pull request by TalentNing**
=== RUN   TestAccAlicloudEssAlarmsDataSource
--- PASS: TestAccAlicloudEssAlarmsDataSource (144.03s)
=== RUN   TestAccAlicloudEssLifecycleHooksDataSource
--- PASS: TestAccAlicloudEssLifecycleHooksDataSource (76.78s)
=== RUN   TestAccAlicloudEssNotificationsDataSource
--- PASS: TestAccAlicloudEssNotificationsDataSource (58.46s)
=== RUN   TestAccAlicloudEssScalingconfigurationsDataSource
--- PASS: TestAccAlicloudEssScalingconfigurationsDataSource (100.20s)
=== RUN   TestAccAlicloudEssScalingGroupsDataSource
--- PASS: TestAccAlicloudEssScalingGroupsDataSource (85.78s)
=== RUN   TestAccAlicloudEssScheduledtasksDataSource
--- PASS: TestAccAlicloudEssScheduledtasksDataSource (108.77s)
=== RUN   TestAccAlicloudEssAlarmBasic
--- PASS: TestAccAlicloudEssAlarmBasic (178.03s)
=== RUN   TestAccAlicloudEssAlarmMulti
--- PASS: TestAccAlicloudEssAlarmMulti (63.98s)
=== RUN   TestAccAlicloudEssAlbServerGroupAttachment_basic
--- PASS: TestAccAlicloudEssAlbServerGroupAttachment_basic (167.13s)
=== RUN   TestAccAlicloudEssEciScalingConfigurationBasic
--- PASS: TestAccAlicloudEssEciScalingConfigurationBasic (51.70s)
=== RUN   TestAccAlicloudEssLifecycleHookBasic
--- PASS: TestAccAlicloudEssLifecycleHookBasic (98.70s)
=== RUN   TestAccAlicloudEssNotification_basic
--- PASS: TestAccAlicloudEssNotification_basic (74.92s)
=== RUN   TestAccAlicloudEssScalingConfigurationUpdate
    provider_test.go:155: Skipping unsupported region cn-hangzhou. Supported regions: [cn-beijing].
--- SKIP: TestAccAlicloudEssScalingConfigurationUpdate (0.00s)
=== RUN   TestAccAlicloudEssScalingConfigurationPerformanceLevel
--- PASS: TestAccAlicloudEssScalingConfigurationPerformanceLevel (108.03s)
=== RUN   TestAccAlicloudEssScalingConfigurationMulti
--- PASS: TestAccAlicloudEssScalingConfigurationMulti (48.20s)
=== RUN   TestAccAlicloudEssScalingGroup_basic
--- PASS: TestAccAlicloudEssScalingGroup_basic (121.59s)
=== RUN   TestAccAlicloudEssScalingGroup_withLaunchTemplateId
--- PASS: TestAccAlicloudEssScalingGroup_withLaunchTemplateId (61.79s)
=== RUN   TestAccAlicloudEssScalingGroup_costoptimized
--- PASS: TestAccAlicloudEssScalingGroup_costoptimized (94.37s)
=== RUN   TestAccAlicloudEssScalingGroup_vpc
--- PASS: TestAccAlicloudEssScalingGroup_vpc (110.04s)
=== RUN   TestAccAlicloudEssScalingGroup_slb
--- PASS: TestAccAlicloudEssScalingGroup_slb (141.41s)
=== RUN   TestAccAlicloudEssScalingGroup_tags
--- PASS: TestAccAlicloudEssScalingGroup_tags (68.38s)
=== RUN   TestAccAlicloudEssVserverGroups_basic
--- PASS: TestAccAlicloudEssVserverGroups_basic (76.90s)
=== RUN   TestAccAlicloudEssScalingRule_basic
--- PASS: TestAccAlicloudEssScalingRule_basic (89.57s)
=== RUN   TestAccAlicloudEssScalingRule_target_tracking_rule_basic
--- PASS: TestAccAlicloudEssScalingRule_target_tracking_rule_basic (97.38s)
=== RUN   TestAccAlicloudEssScalingRule_step_rule_basic
--- PASS: TestAccAlicloudEssScalingRule_step_rule_basic (58.47s)
=== RUN   TestAccAlicloudEssScalingRuleMulti
--- PASS: TestAccAlicloudEssScalingRuleMulti (55.18s)
=== RUN   TestAccAlicloudEssScheduledTask_basic
--- PASS: TestAccAlicloudEssScheduledTask_basic (106.82s)
=== RUN   TestAccAlicloudEssScheduledTask_basic_2
--- PASS: TestAccAlicloudEssScheduledTask_basic_2 (77.67s)
=== RUN   TestAccAlicloudEssScheduledTask_multi
--- PASS: TestAccAlicloudEssScheduledTask_multi (48.64s)

### CODE REVIEW

File: alicloud/resource_alicloud_ess_scaling_group.go [hunk 1]
```diff
@@ -127,6 +128,11 @@ func resourceAlicloudEssScalingGroup() *schema.Resource {
 				Optional: true,
 			},
 			"tags": tagsSchema(),
+			"group_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "ECS",
```
Comment (xinmu0426) [inline alicloud/resource_alicloud_ess_scaling_group.go:134]: 默认值去掉吧，针对新增属性也应该有测试用例

File: website/docs/r/ess_eci_scaling_configuration.html.markdown [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+#### Block container
+
+The container supports the following:
+
+* `args` - (Optional) The arguments passed to the commands.
+* `commands` - (Optional) The commands run by the init container.
+* `cpu` - (Optional) The amount of CPU resources allocated to the container.
+* `environment_vars` - (Optional) The structure of environmentVars.
+    * `key` - (Optional) The name of the variable. The name can be 1 to 128 characters in length and can contain
```
Comment (xinmu0426) [inline website/docs/r/ess_eci_scaling_configuration.html.markdown:193]: 这块也都应该写为Blocks

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceAlicloudEssEciScalingConfiguration() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAliyunEssEciScalingConfigurationCreate,
+		Read:   resourceAliyunEssEciScalingConfigurationRead,
+		Update: resourceAliyunEssEciScalingConfigurationUpdate,
+		Delete: resourceAliyunEssEciScalingConfigurationDelete,
+		Importer: &schema.ResourceImporter{
```
Comment (xinmu0426) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:19]: 增加一个timeout字段配置

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"active": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
```
Comment (xinmu0426) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:26]: Default 值都去掉吧

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 5]
```diff
@@ -0,0 +1,1165 @@
...
+			},
+			"scaling_group_id": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"scaling_configuration_name": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (xinmu0426) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:40]: 加上正则校验，有长度格式要求的都要加上

Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:40]: 暂不提供校验，仅文档说明。

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go (hunk 4 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:26]: done

Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:26]: done

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go (hunk 3 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:19]: done

File: alicloud/resource_alicloud_ess_scaling_group.go (hunk 1 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_scaling_group.go]: done

File: website/docs/r/ess_scaling_group.html.markdown [hunk 6]
```diff
@@ -122,6 +122,7 @@ The following arguments are supported:
 * `group_deletion_protection` - (Optional, Available in v1.102.0+) Specifies whether the scaling group deletion protection is enabled. `true` or `false`, Default value: `false`.            
 * `launch_template_id` - (Optional, Available in v1.141.0+) Instance launch template ID, scaling group obtains launch configuration from instance launch template, see [Launch Template](https://www.alibabacloud.com/help/doc-detail/73916.html). Creating scaling group from launch template enable group automatically.
 * `launch_template_version` - (Optional, Available in v1.159.0+) The version number of the launch template. Valid values are the version number, `Latest`, or `Default`, Default value: `Default`.
+* `group_type` - (Optional, Available in v1.164.0+) Resource type within scaling group. Optional values: ECS, ECI. Default to ECS.
```
Comment (xinmu0426) [inline website/docs/r/ess_scaling_group.html.markdown:125]: Valid values : `ECS`, `ECI`.  Default Value:  `ECS`.

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go (hunk 5 above)
Comment (xinmu0426) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:43]: 为啥不加。scaling_configuration_name  description

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 7]
```diff
@@ -0,0 +1,1165 @@
...
+			}
+		}
+		d.SetPartial("active")
+	}
+
+	request := map[string]interface{}{
+		"ScalingConfigurationId": d.Id(),
+		"RegionId":               client.RegionId,
+	}
+
```
Comment (xinmu0426) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:897]: 加上 update = false这种形式的判断 参考resourceAlicloudAlbRuleUpdate

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 8]
```diff
@@ -0,0 +1,1165 @@
...
+	o, err := essService.DescribeEssEciScalingConfiguration(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	if d.Get("force_delete").(bool) {
```
Comment (xinmu0426) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:1136]: 为啥有两个删除

File: website/docs/r/ess_eci_scaling_configuration.html.markdown (hunk 2 above)
Comment (TalentNing) [reply website/docs/r/ess_eci_scaling_configuration.html.markdown]: done

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go (hunk 5 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go]: done

File: website/docs/r/ess_scaling_group.html.markdown (hunk 6 above)
Comment (TalentNing) [reply website/docs/r/ess_scaling_group.html.markdown:125]: done

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go (hunk 8 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:1145]: 伸缩配置属于伸缩组下资源，伸缩组不允许删除已启用或最后一条伸缩规则，所以这块和原来ess_scaling_configuration资源采用了类似的处理方式，强制删除时，将整个伸缩组删除。

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 9]
```diff
@@ -0,0 +1,1201 @@
...
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+					},
+				},
+			},
+			"volumes": {
+				Type:     schema.TypeList,
+				Optional: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:345]: 这个为啥不是 TypeSet

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 10]
```diff
@@ -0,0 +1,1201 @@
...
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"host_aliases": {
+				Type:     schema.TypeList,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration.go:412]: 还有这个

File: alicloud/resource_alicloud_ess_scaling_group.go [hunk 11]
```diff
@@ -139,31 +145,38 @@ func resourceAliyunEssScalingGroupCreate(d *schema.ResourceData, meta interface{
...
 	client := meta.(*connectivity.AliyunClient)
+	conn, err := client.NewEssClient()
 	essService := EssService{client}
 
+	runtime := util.RuntimeOptions{}
+	runtime.SetAutoretry(true)
 	if err := resource.Retry(5*time.Minute, func() *resource.RetryError {
 		raw, err := client.WithEssClient(func(essClient *ess.Client) (interface{}, error) {
-			return essClient.CreateScalingGroup(request)
+			//return essClient.CreateScalingGroup(request)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_scaling_group.go:155]: 去掉无用的注释

File: alicloud/service_alicloud_ess.go [hunk 12]
```diff
@@ -180,6 +183,35 @@ func (s *EssService) ActivityStateRefreshFunc(activityId string, failStates []st
 	}
 }
 
+func (s *EssService) DescribeEssScalingGroupById(id string) (object map[string]interface{}, err error) {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_ess.go:186]: 这个方法之前应该已经有了吧，为啥还要再写一个

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_ess_scaling_group.go (hunk 11 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_scaling_group.go:155]: done

File: alicloud/service_alicloud_ess.go (hunk 12 above)
Comment (TalentNing) [reply alicloud/service_alicloud_ess.go:186]: 原方式使用sdk调用方式，新方法使用common request方式。由于原方法依赖的较多，改动点较多，因此本次未对引用位置进行替换。

File: alicloud/resource_alicloud_ess_eci_scaling_configuration_test.go [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+									"value": "newValue",
+								},
+							},
+							"working_dir":       "newWorkingDir",
+							"args":              []string{"arg2"},
+							"cpu":               "2",
+							"gpu":               "2",
+							"memory":            "2",
+							"name":              "newName",
+							"image":             "registry-vpc.cn-hangzhou.aliyuncs.com/eci_open/alpine:3.5",
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_eci_scaling_configuration_test.go:352]: 这个image url 中的 cn-hangzhou应该改成 从环境变量中拿

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_ess_eci_scaling_configuration.go [hunk 14]
```diff
@@ -0,0 +1,1201 @@
...
+						},
+						"image": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"image_pull_policy": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"volume_mounts": {
```
[the remark this answers is not in the document]
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration.go:211]: done

File: alicloud/resource_alicloud_ess_eci_scaling_configuration_test.go (hunk 13 above)
Comment (TalentNing) [reply alicloud/resource_alicloud_ess_eci_scaling_configuration_test.go:352]: done

**Pull request merged**